### PR TITLE
6.0

### LINKED ISSUE #678: Upgrade to Popper 2
@FezVrasta and I have been working hard on releasing the next [Popper version (v2)](https://popper.js.org/) since November (the rewrite initially started by him in 2018 though but laid dormant for ages due to difficulties). The rewrite of Popper will be explained more in-depth in an article once the final release is out, but it fixes tons of bugs and API flaws that have come to light over the past 3 years (Popper v1 was released in early 2017 and has been in use since Tippy v0!)

It's a breaking major, so Tippy will need to increment its own major (v6). I'd like to make some other changes to Tippy as well to better integrate with libraries like `react-spring` - effectively stripping the built-in CSS animations out and element creation and making them optional, but of course still easy to use and add in.

### PR DESCRIPTION
**Pull request by atomiks**
Closes #678

### CODE REVIEW

File: MIGRATION_GUIDE.md [hunk 1]
```diff
@@ -1,4 +1,325 @@
...
+anyway, you can set it in `popperOptions`:
+
+```js
+tippy(targets, {
+  popperOptions: {
+    modifiers: [
+      {
+        name: 'preventOverflow',
+        options: {
+          // `boundary` and `rootBoundary` are accepted here
```
Comment (KubaJastrz) [inline MIGRATION_GUIDE.md]: maybe just write it like this:

```js
{
  name: 'preventOverflow',
  options: {
    boundary: 'window'
  }
}
```

`rootBoundary` is new in popper v2 so it's not needed in migration guide.

(I was confused by the comment, so I had to look up the documentation, hence this suggestion)

File: MIGRATION_GUIDE.md [hunk 2]
```diff
@@ -1,4 +1,325 @@
-# Migration Guide: v4 to v5
```
Comment (KubaJastrz) [inline MIGRATION_GUIDE.md]: Maybe we could use `diff` blocks to make migrations easier?

```diff
- old
+ new
```

File: src/template.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+  BACKDROP_CLASS,
+  BOX_CLASS,
+  CONTENT_CLASS,
+  SVG_ARROW_CLASS,
+} from './constants';
+import {div, isElement} from './dom-utils';
+import {Instance, PopperChildren, PopperElement, Props} from './types';
+import {arrayFrom} from './utils';
+
+const innerHTML = (): 'innerHTML' => 'innerHTML';
```
Comment (KubaJastrz) [inline src/template.ts]: Could use an explanation comment. I remember it had something to do with terser, but not sure what exactly

File: src/template.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    box,
+    content: boxChildren.find(node => node.classList.contains(CONTENT_CLASS)),
+    arrow:
+      boxChildren.find(
+        node =>
+          node.classList.contains(ARROW_CLASS) ||
+          node.classList.contains(SVG_ARROW_CLASS),
+      ) || null,
+    backdrop:
+      boxChildren.find(node => node.classList.contains(BACKDROP_CLASS)) || null,
```
Comment (KubaJastrz) [inline src/template.ts]: Does it have to default to `null`? I think `undefined` would be just fine if not found.

File: src/types.ts [hunk 5]
```diff
@@ -1,11 +1,9 @@
-import Popper from 'popper.js';
+import * as Popper from '@popperjs/core';
 
-export type BasePlacement = 'top' | 'bottom' | 'left' | 'right';
+export type BasePlacement = Popper.Placement;
 
 export type Placement = Popper.Placement;
```
Comment (KubaJastrz) [inline src/types.ts]: This is now `Popper.VariationPlacement`, should it change here too?

It seems redundant to have `BasePlacement` and `Placement` equal each other.

File: src/template.ts (hunk 3 above)
Comment (atomiks) [reply src/template.ts]: 🤷‍♂ it's for Firefox extensions. They don't allow `.innerHTML` but aren't smart enough to detect this lol

File: MIGRATION_GUIDE.md (hunk 2 above)
Comment (atomiks) [reply MIGRATION_GUIDE.md]: I think that makes it a little harder to read in some cases, I'm just gonna leave it for now

File: .eslintignore [hunk 6]
```diff
@@ -6,4 +6,7 @@ demo/
 test/
 build/
 rollup.config.js
-rollup.pre-build.js
+rollup.css.js
+babel.config.js
+jest.config.js
+jest-puppeteer.config.js
```
[the remark this answers is not in the document]
Comment (atomiks) [reply .eslintignore]: There was some TypeScript problem or something iirc

**Comment by atomiks** [comment]
I'm going to merge/release this in like 25 hours probably assuming nothing comes up. 

@KubaJastrz if you can, play around with it today 😄

**Comment by KubaJastrz** [comment]
I can't run tippy v6 on codesandbox for some reason :/ But I can run it locally.

https://codesandbox.io/s/awesome-davinci-9tem9

Works just fine for latest v5.

File: website/gatsby-node.js [hunk 7]
```diff
@@ -1,12 +1,30 @@
...
+    'misc',
+    'faq',
+    'motivation',
+  ];
+
+  paths.forEach(path => {
+    createRedirect({
+      fromPath: `/${path}/`,
+      toPath: `/v5/${path}/`,
+      isPermanent: true,
```
Comment (KubaJastrz) [inline website/gatsby-node.js]: Are you sure you want to lock those paths to v5 docs? This way we would have to keep those pages basically forever :P

File: website/patches/@tippy.js+react+3.1.1.patch [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+   var isControlledMode = visible !== undefined;
+   var isSingletonMode = singleton !== undefined;
+@@ -152,7 +150,6 @@ function Tippy(_ref) {
+ 
+   var props = _extends({
+     ignoreAttributes: ignoreAttributes,
+-    multiple: multiple
+   }, restOfNativeProps, {
+     content: component.container
+   });
```
Comment (KubaJastrz) [inline website/patches/@tippy.js+react+3.1.1.patch]: can probably be removed after release

File: website/src/pages/v6/themes.mdx [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+});
+```
+
+<Demo>
+  <Tippy theme="tomato">
+    <Button>Tomato theme</Button>
+  </Tippy>
+</Demo>
+
+> **What is `~=`?**
```
Comment (KubaJastrz) [inline website/src/pages/v6/themes.mdx]: could add a link to MDN for new developers https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors

File: src/createTippy.ts [hunk 10]
```diff
@@ -317,7 +303,7 @@ export default function createTippy(
 
       if (
         instance.state.isVisible &&
-        includes(instance.props.trigger, 'click')
+        instance.props.trigger.indexOf('click') >= 0
```
Comment (KubaJastrz) [inline src/createTippy.ts]: I kinda liked the `includes` helper more :P

File: src/template.ts (hunk 3 above)
Comment (KubaJastrz) [reply src/template.ts]: Then let's write a comment about it.

File: jest.config.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
+module.exports = {
+  preset: 'jest-puppeteer',
+  testMatch: ['<rootDir>/test/**/*.test.js'],
+  globals: {
+    __DEV__: true,
+  },
+  globalSetup: 'jest-environment-puppeteer/setup',
+  globalTeardown: 'jest-environment-puppeteer/teardown',
+  testEnvironment: 'jest-environment-jsdom-fourteen',
```
[the remark this answers is not in the document]
Comment (atomiks) [reply jest.config.js]: Had trouble upgrading. We can upgrade later.

File: src/createTippy.ts (hunk 10 above)
Comment (atomiks) [reply src/createTippy.ts]: Popper 2 inlines `indexOf`, I think it's ok

File: website/gatsby-node.js (hunk 7 above)
Comment (atomiks) [reply website/gatsby-node.js]: The v5 warning messages link to these so unfortunately yes. Every major will now have a `/vN/` prefix for the docs.

**Comment by atomiks** [comment]
> I can't run tippy v6 on codesandbox for some reason :/ But I can run it locally.

I think CodeSandbox uses CJS modules behind the scenes... shouldn't UMD just work anyway though? The `default` export does cause some issues when combined with other named exports... 😕

**Pull request merged**